spec/ordinary-and-exotic-object-behaviours.html

```diff
@@ -0,0 +1,316 @@
+            1. If _numericIndex_ is not *undefined*, then
+              1. Let _current_ be ? TupleGet(_T_, _numericIndex_)
+              1. If _current_ is *undefined* return *false*.
+              1. If IsAccessorDescriptor(_Desc_) is *true*, return *false*.
+              1. If _Desc_.[[Writable]] is present and has value *true*, return *false*.
+              1. If _Desc_.[[Enumerable]] is present and has value *false*, return *false*.
+              1. If _Desc_.[[Configurable]] is present and has value *true*, return *false*.
+              1. If _Desc_.[[Value]] is present, return ! SameValue(_Desc_.[[Value]], _current_).
+          1. Return *false*.
+        </emu-alg>
+      </emu-clause>
+      
+            
+      <emu-clause id="sec-tuple-exotic-objects-get-p-receiver">
+        <h1>[[Get]] ( _P_, _Receiver_ )</h1>
+        <p>When the [[Get]] internal method of a Tuple exotic object _T_ is called with property key _P_ and ECMAScript language value _Receiver_, the following steps are taken:</p>
+        <emu-alg>
+          1. If Type(_P_) is String, then
+            1. Let _numericIndex_ be ! CanonicalNumericIndexString(_P_).
+            1. If _numericIndex_ is not *undefined*, then
+              1. Return ? TupleGet(_T_, _numericIndex_)
+          1. Return *undefined*
+        </emu-alg>
+      </emu-clause>
+
+      <emu-clause id="sec-tuple-exotic-objects-set-p-receiver">
+        <h1>[[Set]] ( _P_, _Receiver_ )</h1>
+        <p>When the [[Set]] internal method of a Tuple exotic object _T_ is called with property key _P_ and ECMAScript language value _Receiver_, the following steps are taken:</p>
+        <emu-alg>
+          1. Return *false*.
+        </emu-alg>
+      </emu-clause>
+      
+      <emu-clause id="sec-tuple-exotic-objects-ownpropertykeys">
+        <h1>[[OwnPropertyKeys]] ( )</h1>
+        <p>When the [[OwnPropertyKeys]] internal method of a Record exotic object _R_ is called, the following steps are taken:</p>
+        <emu-alg>
+          1. Let _keys_ be a new empty List.
+          1. Assert _T_ is Tuple exotic object.
+          1. Let _len_ be _T_.[[TupleLength]].
+          1. For each integer _i_ starting with 0 such that _i_ &lt; _len_ in ascending order, do
+            1. Add ! ToString(_i_) to _keys_.
+          1. Return _keys_
+        </emu-alg>
+      </emu-clause>
+
+      
+      <emu-clause id="sec-tuplecreate" aoid="TupleCreate">
+        <h1>TupleCreate ( _value_ )</h1>
+        <p>This abstract operation TupleCreate takes a _value_ argument. It is used to specify the creation of new Tuple exotic objects. It performs the following steps when called:</p>
+        <emu-alg>
+            1. Assert: Type(_value_) is Tuple.
+            1. Assert: Type(_value_.[[Sequence]]) is List.
+            1. Let _T_ be ! MakeBasicObject(&laquo; [[TupleData]], [[Prototype]] &raquo;).
+            1. Let _T_.[[TupleLength]] be _value_.[[Sequence]].length
```

Comment:
I don't think there's any reason to store the TupleLength like this. Better to just figure it out from _value_.[[Sequence]] when you need it. You could factor that out into an algorithm if you need it--that would be preferable to making a "denormalized" data structure like this. (Also, the length of Lists is not denoted by `.length` but rather "the length of _value_.[[Sequence]]")